### ISSUE TITLE
Hot Reload Full Refresh in v13.4.8

### ISSUE THREAD

**Issue by frankfka** [vercel/next.js#53614]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Next.js v13.4.8 with pages directory routing & Tailwind:

Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:22 PDT 2023; root:xnu-8796.121.3~7/RELEASE_X86_64
    Binaries:
      Node: 16.18.1
      npm: 9.3.1
      Yarn: 1.22.19
      pnpm: 7.5.0
    Relevant Packages:
      next: 13.4.8
      eslint-config-next: 13.4.8
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/frankfka/nextjs-full-refresh-bug

### To Reproduce

See README in https://github.com/frankfka/nextjs-full-refresh-bug

### Describe the Bug

We noticed in our production app that an upgrade from v13.4.7 -> v13.4.8 breaks hot reload in our dialog components (we use [Radix Dialogs](https://www.radix-ui.com/docs/primitives/components/dialog)). Hot reload still works in normal pages, but will trigger a full refresh if a dialog component is edited.

I see a number of updates to hot reload in 13.4.8, perhaps one of them could have triggered this change?

### Expected Behavior

Hot reload works when editing dialog components

<sub>[NEXT-1568](https://linear.app/vercel/issue/NEXT-1568/hot-reload-full-refresh-in-v1348)</sub>

**Comment by jamesedmonston** [comment]
I'm seeing this too. Editing my header and footer components which are children of `_app.tsx` causes the entire page to reload.

**Comment by Tadimsky** [comment]
I also went version by version and found out that it was definitely introduced in this version.
The hash of `_document.js` changes and that's triggering the reload.
I debugger the hot-reloader server and this is where we tell the client to reload the page:
https://github.com/vercel/next.js/blame/93259caac47a820422bb3a37198c115c37613164/packages/next/src/server/dev/hot-reloader.ts#L1223

This is not new, but the client only recently started listening to that message and reloading the page.
That was introduced with this PR (which was part of the 13.4.8 version):
https://github.com/vercel/next.js/pull/51516/files#diff-2b9bd2ade12f637cc8192719277f9a2238c996e2d8ef88f024b773866d32956bR67

@huozhi is this something you could maybe take a look at?

**Comment by pronovaso** [comment]
From version of 13.4.8 I have this behaviour too and is it a little anoing. If I make some changes on dialog componet and in some tabs, I must go all again.

**Comment by jamesedmonston** [comment]
@huozhi I'm happy to supply a repo privately if it's helpful.

**Comment by nathandallaire** [comment]
Hey all, I noticed this several weeks ago and thought I was losing my mind. 

I found this by creating a new next app: `npx create-next-app@latest` (disable app router for this - I didn't test on that). 

If you edit anything within `_app.js` you'll see the page run a full refresh. I even noticed this on components created and imported into `_app.js` - as long as it is rendered within the hierarchy of `_app.js`, it will run a full refresh.

That being said, if there is content rendered through something that is not `_app.js` (ie. through `pages/index.js`), then it will hot reload as usual. I can confirm I triple checked all the components are named (not anonymous) and their corresponding files are pascal cased.
